gnd/src/main/java/com/google/android/gnd/persistence/local/room/RoomLocalDataStore.java

```diff
@@ -424,8 +427,25 @@ public Completable apply(ObservationMutation mutation) throws LocalDataStoreExce
         return getUser(mutation.getUserId())
             .flatMapCompletable(user -> updateObservation(mutation, user));
       case DELETE:
-        return getUser(mutation.getUserId())
-            .flatMapCompletable(user -> markObservationDeleted(mutation));
+        return observationDao
+            .findById(mutation.getObservationId())
+            .flatMapCompletable(
+                entity -> {
+                  // Mark the ObservationEntity as DELETED if the current state is DEFAULT.
+                  // Otherwise, delete the observation entity from local db.
+                  //
+                  // TODO: If the remote sync fails, reset the state to DEFAULT.
+                  //
+                  // Note: Observations marked as DELETED are not shown in UI.
+                  //       See {@link ObservationDao}
+                  if (entity.getState() == EntityState.DEFAULT) {
```

Comment:
This logic is not idempotent, which can lead to issues wrt order and timing of mutations; if you call it once, it will do something different than if you call it twice. 

It's not clear to me when/how the second mutation to delete the entity would occur. Wouldn't it occur at a lower level, once the remote observation is actually removed (i.e. not via the mutation interface)?

Reply:
I may have misinterpreted your previous comment. You mentioned that `deleteObservation` is another case of apply and should be called via it rather than directly. So I merged the two methods `markObservationDeleted` and `deleteObservation` under the same `apply` implementation.

Current flow:
mark observation as deleted -> create mutation -> remote sync (delete) -> delete local observation -> delete mutation

@gino-m Did I make an error in understanding the comments? Please confirm.

Reply:
Sorry, I may have misspoke - specifically "mark deleted" is a special case of `apply`. "Delete permanently" isn't represented as a mutation type, since it's not a user action that we need to sync to/from the remote server.  

The flow as you described it seems correct. My comment was relating to how the "delete local observation" step is being invoked - it currently depends on how many times this method is called, and as mentioned, the cleanup step isn't a synced mutation type. Instead, it can be done directly in the db after the remote observation is deleted.

Reply:
Done 3077b29